compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp

```diff
@@ -0,0 +1,217 @@
+            }
+         else
+            {
+            callInstr = generateImmSymInstruction(CALLImm4, data.callNode, 0, data.methodSymRef, data.regDeps, data.cg);
+            }
+         }
+      else
+         {
+         /**
+          * Ideally, only the RegImm64Sym instruction form would be used.  However, there are
+          * subtleties in the handling of relocations between these different forms of instructions
+          * that need to be sorted out first..
+          */
+         if (data.useSymInstruction)
+            {
+            TR::AMD64RegImm64SymInstruction *loadInstr;
+
+            if (data.prevInstr)
+               {
+               loadInstr = generateRegImm64SymInstruction(
+                  data.prevInstr,
+                  MOV8RegImm64,
+                  data.scratchReg,
+                  targetAddress,
+                  data.methodSymRef,
+                  data.cg);
+               }
+            else
+               {
+               loadInstr = generateRegImm64SymInstruction(
+                  MOV8RegImm64,
+                  data.callNode,
+                  data.scratchReg,
+                  targetAddress,
+                  data.methodSymRef,
+                  data.cg);
+               }
+
+            if (data.reloKind != TR_NoRelocation)
+               {
+               loadInstr->setReloKind(data.reloKind);
+               }
+
+            data.out_loadInstr = loadInstr;
+            }
+         else
+            {
+            TR::AMD64RegImm64Instruction *loadInstr;
+
+            if (data.prevInstr)
+               {
+               loadInstr = generateRegImm64Instruction(
+                  data.prevInstr,
+                  MOV8RegImm64,
+                  data.scratchReg,
+                  targetAddress,
+                  data.cg,
+                  data.reloKind);
+               }
+            else
+               {
+               loadInstr = generateRegImm64Instruction(
+                  MOV8RegImm64,
+                  data.callNode,
+                  data.scratchReg,
+                  targetAddress,
+                  data.cg,
+                  data.reloKind);
+               }
+
+            data.out_loadInstr = loadInstr;
+            }
+
+         TR_X86OpCodes op = data.isCall ? CALLReg : JMPReg;
+         if (data.prevInstr)
+            {
+            callInstr = generateRegInstruction(data.out_loadInstr, op, data.scratchReg, data.regDeps, data.cg);
+            }
+         else
+            {
+            callInstr = generateRegInstruction(op, data.callNode, data.scratchReg, data.regDeps, data.cg);
+            }
+         }
+      }
+
+   data.out_callInstr = callInstr;
+
+   return callInstr;
+   }
+
+
+uint8_t *
+OMR::X86::AMD64::JitCodeRWXObjectFormat::encodeFunctionCall(TR::FunctionCallData &data)
+   {
+   uint8_t op = data.isCall ? 0xe8 : 0xe9;   // CALL rel32 | JMP rel32
+   *data.bufferAddress++ = op;
```

Comment:
Personally, I find this kind of combined dereference and increment a bit difficult to read.

Reply:
Fair enough.